# Новые фичи
Цифры по mcc  
Погода по месту  
расстояние до дальнейшего соседа  
максимальная продолжительность приобретений в данной точке по дням

ПРОССУМИРОВАТЬ ДЕЛЬТЫ ПО РАЗНЫМ КООРДИНАТАМ

# [Boosters] Raiffeisen Data Cup. Baseline
Общий подход:
- Добавляем к каждой транзакции столбец: is_work (если транзакция находится в пределах 0.02 от дома клиента)
- Добавляем к каждой транзакции столбец: is_home (если транзакция находится в пределах 0.02 от работы клиента)
- Обучаем классификатор предсказывающий вероятность (is_home == 1) для транзакции
- Обучаем классификатор предсказывающий вероятность (is_work == 1) для транзакции

Точность определения местоположения:
- для классификатора is_home: ~3x%
- для классификатора is_work: ~2x%
- общая оценка на Public Leaderboard: ???

Примечание
* Требуется Python версии 3.5
* Требуется библиотека xgboost (для обучения использовалась xgboost версии 0.7.post3)
* Требуются файлы: test_set.csv, train_set.csv в одном каталоге с данным скриптом
* Требования к памяти: должно работать с 2Гб свободного RAM
* Время работы: ~3 минуты (тестировалось на процессоре Intel Core i7-4770)

In [1]:
%load_ext autoreload
%autoreload 2

import sys
MODULES_PATH = '../code/'
if MODULES_PATH not in sys.path:
    sys.path.append(MODULES_PATH)
import mfuncs
    
import pandas as pd
import numpy as np
from tqdm import tqdm
tqdm.pandas()
pd.options.display.max_columns = 1000
pd.options.display.max_colwidth = -1

import lightgbm as lgb


from sklearn.neighbors import NearestNeighbors
from sklearn.cluster import KMeans, MeanShift, estimate_bandwidth, AgglomerativeClustering
from sklearn.metrics import silhouette_samples, silhouette_score

from sklearn.metrics.pairwise import pairwise_distances
import gmaps
API_KEY = 'AIzaSyCG_RL0_kavuEaJAqEN5xXbU4h0VJUbA9M'
gmaps.configure(api_key=API_KEY) # Your Google API key
%pylab inline
figsize(13, 13)

Populating the interactive namespace from numpy and matplotlib


In [2]:
# Определим типы колонок для экономии памяти
dtypes = pd.read_csv('../data/df_all_b11_dtypes.csv', header=None, index_col=0).to_dict()[1]
dtypes.pop('transaction_date', None)
df_all = pd.read_csv('../data/df_all_b11.csv', dtype=dtypes, parse_dates=['transaction_date'])

## Мои фичи

In [3]:
# добавим признаки после групбая
df_gb = df_all[['customer_id','amount', 'add_lat', 'add_lon']].groupby('customer_id')
coord_stat_df = df_gb.agg(['mean', 'max', 'min'])
coord_stat_df['transactions_per_user'] = df_gb.agg('size')
coord_stat_df.columns = ['_'.join(col).strip() for col in coord_stat_df.columns.values]
coord_stat_df = coord_stat_df.astype(np.float32)
coord_stat_df.reset_index(inplace=True)
df_all = pd.merge(df_all, coord_stat_df, on='customer_id', how='left')

In [4]:
cols = ['add_lat', 'add_lon']
types = ['min', 'max', 'mean']
for c in cols:
    for t in types:
        df_all['{}_diff_{}'.format(c, t)] = np.abs(df_all[c] - df_all['{}_{}'.format(c, t)], dtype=np.float32)

In [5]:
df_all = df_all.loc[:,~df_all.columns.duplicated()]

In [6]:
# разности 
df_all['lat_diff_cluster_lat'] = np.abs(df_all['add_lat'] - df_all['cl_lat'], dtype=np.float32)
df_all['lon_diff_cluster_lon'] = np.abs(df_all['add_lon'] - df_all['cl_lon'], dtype=np.float32)
df_all['lon_diff_cluster'] = (df_all['lat_diff_cluster_lat'] + df_all['lon_diff_cluster_lon']).astype(np.float32)

## Категории mcc

In [7]:
# категории
df_all['mcc_str'] = df_all['mcc'].astype(str).str.rjust(4, '0')
df_mcc = pd.read_csv('../data/internal/mcc.csv')
df_mcc = df_mcc.iloc[1:, :3]
df_mcc.columns = ['mcc_str', 'mcc_cat1', 'mcc_cat2']
df_mcc.drop_duplicates(subset=['mcc_str'], inplace=True)
df_mcc['mcc_cat1'] = pd.factorize(df_mcc['mcc_cat1'])[0].astype(np.int32)
df_mcc['mcc_cat2'] = pd.factorize(df_mcc['mcc_cat2'])[0].astype(np.int32)
df_mcc.fillna('none', inplace=True)
df_all = pd.merge(df_all, df_mcc, on='mcc_str', how='left')
del df_all['mcc_str']
df_mcc.head()

mcc_str  mcc_cat1  mcc_cat2
1  0001   -1         0       
2  0002   -1         0       
3  0003   -1         0       
4  0004   -1         0       
5  0005   -1         0

## Плотности населения по районам МСК

In [8]:
import geopandas as gpd
from shapely.geometry import Point, Polygon
mos_shp = gpd.read_file('../data/internal/demography.shp')

_pnts = [Point(vals.T) for vals in df_all[df_all.city_name=='Москва'][['add_lon', 'add_lat']].values]
pnts = gpd.GeoDataFrame(geometry=_pnts)
pnts.crs = mos_shp.crs

mos_shp.drop(['NAME', 'ABBREV_AO'], 1, inplace=True)
mos_shp['area'] = mos_shp['geometry'].area
for c in mos_shp.columns:
    if c not in ['geometry', 'area'] and 'index' not in c:
        mos_shp[c + 'dens'] = mos_shp[c] / mos_shp['area']

In [9]:
%%time
cities_with_country = gpd.sjoin(pnts, mos_shp, how="left", op='intersects')

CPU times: user 44 s, sys: 260 ms, total: 44.3 s
Wall time: 44.3 s


In [10]:
cols = cities_with_country.drop(['geometry', 'index_right'], 1).columns
for c in cols:
    df_all[c] = -1
df_all.loc[df_all.city_name=='Москва', cols] = cities_with_country

In [11]:
# частота mcc
df_mcc = df_all['mcc'].value_counts(normalize=True).reset_index()
df_mcc.columns = ['mcc', 'mcc_freq']
df_all = pd.merge(df_all, df_mcc, on='mcc', how='left')

In [12]:
# метро
mos_metro = pd.read_csv('../data/internal/moscow_metro.csv')
pet_metro = pd.read_csv('../data/internal/peter_metro.csv')
df_metro = pd.concat([mos_metro, pet_metro])

vals1 = df_all[['add_lat', 'add_lon']].values
vals2 = df_metro[['metro_lat', 'metro_lon']].values
X = pairwise_distances(vals1, vals2)
dist_to_min_metro = X.min(axis=1)

# X[X == 0] = 10000
df_all['dist_to_minmetro'] = X.min(axis=1)
df_all['metro_in_01'] = (X < 0.01).sum(axis=1)
df_all['metro_in_001'] = (X < 0.001).sum(axis=1)
df_all['metro_in_02'] = (X < 0.02).sum(axis=1)
df_all['metro_in_005'] = (X < 0.005).sum(axis=1)
df_all['metro_in_03'] = (X < 0.03).sum(axis=1)

In [13]:
# расстояние до участковых комиссий
df_cik = pd.read_csv('../data/internal/cik_uik.csv')
df_cik.dropna(subset=['lat_ik'], inplace=True)
df_cik.dropna(subset=['lon_ik'], inplace=True)

df_cik = df_cik[df_cik['lon_ik'] < 45]
vals1 = df_all[['add_lat', 'add_lon']].drop_duplicates().values.astype(np.float32)
df_vals = pd.DataFrame(vals1, columns=['add_lat', 'add_lon'])
vals2 = df_cik[['lat_ik', 'lon_ik']].drop_duplicates().values.astype(np.float32)

print(vals2.shape)
X = pairwise_distances(vals1, vals2)

df_vals['dist_to_ciktro'] = X.min(axis=1)
df_vals['cik_in_01'] = (X < 0.01).sum(axis=1)
df_vals['cik_in_001'] = (X < 0.001).sum(axis=1)
df_vals['cik_in_02'] = (X < 0.02).sum(axis=1)
df_vals['cik_in_005'] = (X < 0.005).sum(axis=1)
df_vals['cik_in_03'] = (X < 0.03).sum(axis=1)

(37481, 2)


In [14]:
df_all['add_lat_'] = np.round(df_all['add_lat'] * 10000).astype(int) 
df_all['add_lon_'] = np.round(df_all['add_lon'] * 10000).astype(int) 
df_vals['add_lat_'] = np.round(df_vals['add_lat'] * 10000).astype(int) 
df_vals['add_lon_'] = np.round(df_vals['add_lon'] * 10000).astype(int)
df_vals.drop_duplicates(subset=['add_lat_', 'add_lon_'], inplace=True)
del df_vals['add_lat']
del df_vals['add_lon']

df_all = pd.merge(df_all, df_vals, on=['add_lat_', 'add_lon_'], how='left')
del X
del df_all['add_lat_']
del df_all['add_lon_']

In [15]:
# погода в МСК и ПИТЕРЕ
# буду смотреть погоду в 18-00
w1 = pd.read_csv('../data/internal/weather/moscow.csv', sep=';', index_col=False)
w1['city_name'] = 'Москва'
w1['transaction_date'] = pd.to_datetime(w1['Local time in Moscow'], format='%d.%m.%Y %H:%M')
del w1['Local time in Moscow']
w1 = w1[w1.transaction_date.dt.hour == 18].reset_index(drop=True)
w1['transaction_date'] = w1['transaction_date'].dt.date

w2 = pd.read_csv('../data/internal/weather/peter.csv', sep=';', index_col=False)
w2['city_name'] = 'Санкт-Петербург'
w2['transaction_date'] = pd.to_datetime(w2['Local time in Moscow'], format='%d.%m.%Y %H:%M')
del w2['Local time in Moscow']
w2 = w2[w2.transaction_date.dt.hour == 18].reset_index(drop=True)
w2['transaction_date'] = w2['transaction_date'].dt.date

df_weather = pd.concat([w1, w2], axis=0).reset_index(drop=True)
df_weather['transaction_date'] = pd.to_datetime(df_weather['transaction_date'])

cn = df_weather['city_name'] # hardcode
df_weather = df_weather.select_dtypes(exclude=['object'])
df_weather['city_name'] = cn 
for c in df_weather:
    if df_weather[c].isnull().mean() > 0.9:
        del df_weather[c]
# df_weather = df_weather.add_prefix('weather_')
df_all = pd.merge(df_all, df_weather, on=['city_name', 'transaction_date'], how='left')

In [19]:
# df_all.drop(['index', 'T', 'Po', 'P', 'Pa', 'U', 'Ff', 'VV', 'Td', 'tR'], 1, inplace=True)

In [20]:
# добавляем новые MCC OHE с самыми частыми категориями
df_all['mcc_rm'] = df_all['mcc']
df_all.loc[~df_all['mcc_rm'].isin(df_all['mcc_rm'].value_counts().iloc[:35].index.values), 'mcc_rm'] = 99999

df_all['mcc_rm_cat1'] = df_all['mcc_cat1']
df_all.loc[~df_all['mcc_rm_cat1'].isin(df_all['mcc_rm_cat1'].value_counts().iloc[:35].index.values), 
           'mcc_rm_cat1'] = 99999

In [21]:
# OHE урезанных MCC
df_all = pd.concat([df_all, 
                    pd.get_dummies(df_all['mcc_rm'], prefix='mcc_rm_ohe').astype(np.int8)], axis=1)
del df_all['mcc_rm']
df_all = pd.concat([df_all, 
                    pd.get_dummies(df_all['mcc_rm_cat1'], prefix='mcc_rm_cat1_ohe').astype(np.int8)], axis=1)
del df_all['mcc_rm_cat1']

df_all = pd.concat([df_all, 
                    pd.get_dummies(df_all['mcc_cat2'], prefix='mcc_cat2_ohe').astype(np.int8)], axis=1)
del df_all['mcc_cat2']
df_all = df_all.reset_index(drop=True)

In [22]:
mcc_cols_0 = [c for c in df_all.columns if 'mcc_rm_ohe' in c]
mcc_cols_1 = [c for c in df_all.columns if 'mcc_rm_cat1_ohe' in c]
mcc_cols_2 = [c for c in df_all.columns if 'mcc_cat2_ohe' in c]


mcc_cols_0_ = [c + '_amount' for c in mcc_cols_0]
mcc_cols_1_ = [c + '_amount' for c in mcc_cols_1]
mcc_cols_2_ = [c + '_amount' for c in mcc_cols_2]

In [23]:
# сделаем групбай какие вообще есть mcc у посетителя. Это поможет понять его привычки
df_mcc = df_all.groupby('customer_id')[mcc_cols_0].agg(['mean', 'sum'])
df_mcc.columns = ['_'.join(col).strip() for col in df_mcc.columns.values]
df_mcc = df_mcc.astype(np.float32).reset_index()
df_all = pd.merge(df_all, df_mcc, on='customer_id', how='left')
df_mcc.head()

customer_id  mcc_rm_ohe_4111_mean  \
0  0001f322716470bf9bfc1708f06f00fc  0.000000               
1  00021683ccb416637fe9a4cd35e4606e  0.000000               
2  0002d0f8a642272b41c292c12ab6e602  0.000000               
3  0004d182d9fede3ba2534b2d5e5ad27e  0.000000               
4  0007297d86e14bd68bd87b1dbdefe302  0.008333               

   mcc_rm_ohe_4111_sum  mcc_rm_ohe_4784_mean  mcc_rm_ohe_4784_sum  \
0  0.0                  0.0                   0.0                   
1  0.0                  0.0                   0.0                   
2  0.0                  0.0                   0.0                   
3  0.0                  0.0                   0.0                   
4  2.0                  0.0                   0.0                   

   mcc_rm_ohe_5200_mean  mcc_rm_ohe_5200_sum  mcc_rm_ohe_5211_mean  \
0  0.0                   0.0                  0.02                   
1  0.0                   0.0                  0.00                   
2  0.0                   0.0                  0.00                   
3  0.0                   0.0                  0.00                   
4  0.0                   0.0                  0.00                   

   mcc_rm_ohe_5211_sum  mcc_rm_ohe_5261_mean  mcc_rm_ohe_5261_sum  \
0  1.0                  0.000000              0.0                   
1  0.0                  0.038462              3.0                   
2  0.0                  0.000000              0.0                   
3  0.0                  0.000000              0.0                   
4  0.0                  0.000000              0.0                   

   mcc_rm_ohe_5311_mean  mcc_rm_ohe_5311_sum  mcc_rm_ohe_5331_mean  \
0  0.000000              0.0                  0.000000               
1  0.000000              0.0                  0.012821               
2  0.051948              4.0                  0.000000               
3  0.008333              1.0                  0.000000               
4  0.000000              0.0                  0.000000               

   mcc_rm_ohe_5331_sum  mcc_rm_ohe_5411_mean  mcc_rm_ohe_5411_sum  \
0  0.0                  0.340000              17.0                  
1  1.0                  0.371795              29.0                  
2  0.0                  0.610390              47.0                  
3  0.0                  0.150000              18.0                  
4  0.0                  0.270833              65.0                  

   mcc_rm_ohe_5499_mean  mcc_rm_ohe_5499_sum  mcc_rm_ohe_5533_mean  \
0  0.020000              1.0                  0.0                    
1  0.012821              1.0                  0.0                    
2  0.000000              0.0                  0.0                    
3  0.016667              2.0                  0.0                    
4  0.008333              2.0                  0.0                    

   mcc_rm_ohe_5533_sum  mcc_rm_ohe_5541_mean  mcc_rm_ohe_5541_sum  \
0  0.0                  0.140000              7.0                   
1  0.0                  0.000000              0.0                   
2  0.0                  0.000000              0.0                   
3  0.0                  0.000000              0.0                   
4  0.0                  0.004167              1.0                   

   mcc_rm_ohe_5641_mean  mcc_rm_ohe_5641_sum  mcc_rm_ohe_5651_mean  \
0  0.000000              0.0                  0.0                    
1  0.012821              1.0                  0.0                    
2  0.000000              0.0                  0.0                    
3  0.000000              0.0                  0.0                    
4  0.000000              0.0                  0.0                    

   mcc_rm_ohe_5651_sum  mcc_rm_ohe_5661_mean  mcc_rm_ohe_5661_sum  \
0  0.0                  0.000000              0.0                   
1  0.0                  0.025641              2.0                   
2  0.0                  0.000000              0.0                   
3  0.0                  0.016667     

In [24]:
# по объемам
for i, c in enumerate(mcc_cols_0):
    df_all[mcc_cols_0_[i]] = (df_all[c] * df_all['amount']).astype(np.float32)
for i, c in enumerate(mcc_cols_2):
    df_all[mcc_cols_2_[i]] = (df_all[c] * df_all['amount']).astype(np.float32)

In [25]:
# по объемам
df_mcc = df_all.groupby('customer_id')[mcc_cols_0_].agg(['mean', 'sum'])
df_mcc.columns = ['_'.join(col).strip() for col in df_mcc.columns.values]
df_mcc = df_mcc.astype(np.float32).reset_index()
df_all = pd.merge(df_all, df_mcc, on='customer_id', how='left')
df_mcc.head()

# df_all['add_lat_'] = (df_all['add_lat'] * 40).astype(np.int32)
# df_all['add_lon_'] = (df_all['add_lon'] * 40).astype(np.int32)

# df_mcc = df_all.groupby(['add_lat_', 'add_lon_'])[mcc_cols_].agg(['mean', 'sum'])
# df_mcc = df_mcc.add_suffix('_40coord')
# df_mcc.columns = ['_'.join(col).strip() for col in df_mcc.columns.values]
# df_mcc = df_mcc.astype(np.float32)
# df_mcc.reset_index(inplace=True)
# df_mcc.head()
# df_all = pd.merge(df_all, df_mcc, on=['add_lat_', 'add_lon_'], how='left')

# del df_all['add_lat_']
# del df_all['add_lon_']

customer_id  mcc_rm_ohe_4111_amount_mean  \
0  0001f322716470bf9bfc1708f06f00fc  0.000000                      
1  00021683ccb416637fe9a4cd35e4606e  0.000000                      
2  0002d0f8a642272b41c292c12ab6e602  0.000000                      
3  0004d182d9fede3ba2534b2d5e5ad27e  0.000000                      
4  0007297d86e14bd68bd87b1dbdefe302  4.562878                      

   mcc_rm_ohe_4111_amount_sum  mcc_rm_ohe_4784_amount_mean  \
0  0.00000                     0.0                           
1  0.00000                     0.0                           
2  0.00000                     0.0                           
3  0.00000                     0.0                           
4  1095.09082                  0.0                           

   mcc_rm_ohe_4784_amount_sum  mcc_rm_ohe_5200_amount_mean  \
0  0.0                         0.0                           
1  0.0                         0.0                           
2  0.0                         0.0                           
3  0.0                         0.0                           
4  0.0                         0.0                           

   mcc_rm_ohe_5200_amount_sum  mcc_rm_ohe_5211_amount_mean  \
0  0.0                         27.80541                      
1  0.0                         0.00000                       
2  0.0                         0.00000                       
3  0.0                         0.00000                       
4  0.0                         0.00000                       

   mcc_rm_ohe_5211_amount_sum  mcc_rm_ohe_5261_amount_mean  \
0  1390.270508                 0.000000                      
1  0.000000                    75.903763                     
2  0.000000                    0.000000                      
3  0.000000                    0.000000                      
4  0.000000                    0.000000                      

   mcc_rm_ohe_5261_amount_sum  mcc_rm_ohe_5311_amount_mean  \
0  0.000000                    0.000000                      
1  5920.493164                 0.000000                      
2  0.000000                    8.041179                      
3  0.000000                    3.977230                      
4  0.000000                    0.000000                      

   mcc_rm_ohe_5311_amount_sum  mcc_rm_ohe_5331_amount_mean  \
0  0.000000                    0.000000                      
1  0.000000                    46.309711                     
2  619.170776                  0.000000                      
3  477.267609                  0.000000                      
4  0.000000                    0.000000                      

   mcc_rm_ohe_5331_amount_sum  mcc_rm_ohe_5411_amount_mean  \
0  0.000000                    174.479874                    
1  3612.157471                 447.340057                    
2  0.000000                    102.385597                    
3  0.000000                    82.209663                     
4  0.000000                    237.856369                    

   mcc_rm_ohe_5411_amount_sum  mcc_rm_ohe_5499_amount_mean  \
0  8723.993164                 3.240682                      
1  34892.523438                2.944012                      
2  7883.690918                 0.000000                      
3  9865.159180                 3.002061                      
4  57085.527344                0.716898                      

   mcc_rm_ohe_5499_amount_sum  mcc_rm_ohe_5533_amount_mean  \
0  162.034088                  0.0                           
1  229.632919                  0.0                           
2  0.000000                    0.0                           
3  360.247314                  0.0                           
4  172.055511                  0.0                           

   mcc_rm_ohe_5533_amount_sum  mcc_rm_ohe_5541_amount_mean  \
0  0.0                         142.566284                    
1  0.0                         0.000000                      
2  0.0                         0.000000                      
3  0.0    

In [26]:
# по объемам
df_mcc = df_all.groupby('customer_id')[mcc_cols_2_].agg(['mean', 'sum'])
df_mcc.columns = ['_'.join(col).strip() for col in df_mcc.columns.values]
df_mcc = df_mcc.astype(np.float32).reset_index()
df_mcc.head()
df_all = pd.merge(df_all, df_mcc, on='customer_id', how='left')

# df_all['add_lat_'] = (df_all['add_lat'] * 40).astype(np.int32)
# df_all['add_lon_'] = (df_all['add_lon'] * 40).astype(np.int32)

# df_mcc = df_all.groupby(['add_lat_', 'add_lon_'])[mcc_cols_].agg(['mean', 'sum'])
# df_mcc = df_mcc.add_suffix('_40coord')
# df_mcc.columns = ['_'.join(col).strip() for col in df_mcc.columns.values]
# df_mcc = df_mcc.astype(np.float32)
# df_mcc.reset_index(inplace=True)
# df_mcc.head()
# df_all = pd.merge(df_all, df_mcc, on=['add_lat_', 'add_lon_'], how='left')

# del df_all['add_lat_']
# del df_all['add_lon_']

In [27]:
# сделаем групбай какие вообще есть mcc у посетителя. Это поможет понять его привычки
# mcc_cols = [c for c in df_all.columns if 'mcc_cat1' in c]
# df_mcc = df_all.groupby('customer_id')[mcc_cols].agg(['mean'])
# df_mcc.columns = ['_'.join(col).strip() for col in df_mcc.columns.values]
# df_mcc.reset_index(inplace=True)
# df_mcc.head()
# df_all = pd.merge(df_all, df_mcc, on='customer_id', how='left')

In [28]:
# сделаем групбай какие вообще есть mcc у посетителя. Это поможет понять его привычки
df_mcc = df_all.groupby('customer_id')[mcc_cols_2].agg(['mean', 'sum'])
df_mcc.columns = ['_'.join(col).strip() for col in df_mcc.columns.values]
df_mcc = df_mcc.astype(np.float32)
df_mcc.reset_index(inplace=True)
df_mcc.head()
df_all = pd.merge(df_all, df_mcc, on='customer_id', how='left')

In [29]:
# РАСПРЕДЕЛЕНИЕ MCC В ОКРЕСТНОСТИ ЧУВАКА
df_all['add_lat_'] = (df_all['add_lat'] * 40).astype(np.int32)
df_all['add_lon_'] = (df_all['add_lon'] * 40).astype(np.int32)

df_mcc = df_all.groupby(['add_lat_', 'add_lon_'])[mcc_cols_0].agg(['mean', 'sum'])
df_mcc = df_mcc.add_suffix('_40coord')
df_mcc.columns = ['_'.join(col).strip() for col in df_mcc.columns.values]
df_mcc = df_mcc.astype(np.float32)
df_mcc.reset_index(inplace=True)
df_mcc.head()
df_all = pd.merge(df_all, df_mcc, on=['add_lat_', 'add_lon_'], how='left')

del df_all['add_lat_']
del df_all['add_lon_']

In [30]:
mcc_cols = [c for c in df_all.columns if 'mcc_rm_ohe' in c and 'mean' not in c and 'sum' not in c]
# РАСПРЕДЕЛЕНИЕ MCC В ОКРЕСТНОСТИ ЧУВАКА
df_all['add_lat_'] = (df_all['add_lat'] * 100).astype(np.int32)
df_all['add_lon_'] = (df_all['add_lon'] * 100).astype(np.int32)

df_mcc = df_all.groupby(['add_lat_', 'add_lon_'])[mcc_cols_0].agg(['mean', 'sum'])
df_mcc = df_mcc.add_suffix('_100coord')
df_mcc.columns = ['_'.join(col).strip() for col in df_mcc.columns.values]
df_mcc = df_mcc.astype(np.float32)
df_mcc.reset_index(inplace=True)
df_mcc.head()
df_all = pd.merge(df_all, df_mcc, on=['add_lat_', 'add_lon_'], how='left')

del df_all['add_lat_']
del df_all['add_lon_']

In [31]:
# РАСПРЕДЕЛЕНИЕ MCC В ОКРЕСТНОСТИ ЧУВАКА (ПРОВЕРИЛ-ЛУЧШЕ РАБОТАЕТ НА БОЛЬШИХ УЧАСТКАХ)
df_all['add_lat_'] = (df_all['add_lat'] * 40).astype(np.int32)
df_all['add_lon_'] = (df_all['add_lon'] * 40).astype(np.int32)

df_mcc = df_all.groupby(['add_lat_', 'add_lon_'])[mcc_cols_2].agg(['mean', 'sum'])
df_mcc = df_mcc.add_suffix('_200coord')
df_mcc.columns = ['_'.join(col).strip() for col in df_mcc.columns.values]
df_mcc = df_mcc.astype(np.float32)
df_mcc.reset_index(inplace=True)
df_mcc.head()
df_all = pd.merge(df_all, df_mcc, on=['add_lat_', 'add_lon_'], how='left')

del df_all['add_lat_']
del df_all['add_lon_']

In [32]:
# РАСПРЕДЕЛЕНИЕ MCC В ОКРЕСТНОСТИ ЧУВАКА
# df_all['add_lat_'] = (df_all['add_lat'] * 100).astype(np.int32)
# df_all['add_lon_'] = (df_all['add_lon'] * 100).astype(np.int32)

# df_mcc = df_all.groupby(['add_lat_', 'add_lon_'])[mcc_cols].agg(['mean', 'sum'])
# df_mcc = df_mcc.add_suffix('_100coord')
# df_mcc.columns = ['_'.join(col).strip() for col in df_mcc.columns.values]
# df_mcc = df_mcc.astype(np.float32)
# df_mcc.reset_index(inplace=True)
# df_mcc.head()
# df_all = pd.merge(df_all, df_mcc, on=['add_lat_', 'add_lon_'], how='left')

# del df_all['add_lat_']
# del df_all['add_lon_']

# Игрушки с адресами

In [33]:
df_all['string'] = df_all['string'].fillna('')
df_all['string'] = df_all['string'].str.lower()

In [34]:
df_all['has_street'] = df_all['string'].str.contains('улиц').astype(np.int8)
df_all['has_pereul'] = df_all['string'].str.contains('переул').astype(np.int8)
df_all['has_bulvar'] = df_all['string'].str.contains('бульв').astype(np.int8)
df_all['has_prospekt'] = df_all['string'].str.contains('проспект').astype(np.int8)
df_all['has_shosse'] = df_all['string'].str.contains('шосс').astype(np.int8)

df_all['has_torg'] = df_all['string'].str.contains('торгов').astype(np.int8)
df_all['has_bus'] = df_all['string'].str.contains('бизн').astype(np.int8)

# Медианы по юзеру и по без дубликатов

In [35]:
dft = df_all.groupby('terminal_id')['add_lat'].agg('std').astype(np.float32).reset_index()
dft['moving_terminal'] = (dft['add_lat'] > 0).astype(np.int8)
del dft['add_lat']
df_all = pd.merge(df_all, dft, on='terminal_id', how='left')

In [36]:
df_med = df_all.groupby('customer_id')['add_lat', 'add_lon'].agg('median').astype(np.float32).reset_index()
df_med.columns = ['customer_id', 'add_lat_median', 'add_lon_median']
df_all = pd.merge(df_all, df_med, on='customer_id', how='left')

In [37]:
df_med = df_all.drop_duplicates(subset=['customer_id', 
                                        'add_lat', 'add_lon']).groupby('customer_id')['add_lat', 'add_lon'].agg('median').reset_index()
df_med.columns = ['customer_id', 'add_lat_median_unique', 'add_lon_median_unique']
df_all = pd.merge(df_all, df_med, on='customer_id', how='left')

In [38]:
df_all['lat_diff_median'] = np.abs(df_all['add_lat'] - df_all['add_lat_median'])
df_all['lon_diff_median'] = np.abs(df_all['add_lon'] - df_all['add_lat_median'])
df_all['lat_diff_median_unique'] = np.abs(df_all['add_lat'] - df_all['add_lat_median_unique'])
df_all['lon_diff_median_unique'] = np.abs(df_all['add_lon'] - df_all['add_lon_median_unique'])

df_all['diff_median'] = df_all['lat_diff_median'] + df_all['lon_diff_median']
df_all['diff_median_unique'] = df_all['lat_diff_median_unique'] + df_all['lon_diff_median_unique']

In [39]:
del dft
del df_med

# OSM
https://wiki.openstreetmap.org/wiki/RU:%D0%9E%D0%B1%D1%8A%D0%B5%D0%BA%D1%82%D1%8B_%D0%BA%D0%B0%D1%80%D1%82%D1%8B#.D0.9A.D0.BE.D0.BC.D0.BC.D0.B5.D1.80.D1.87.D0.B5.D1.81.D0.BA.D0.B8.D0.B5

In [40]:
import ogr
driver=ogr.GetDriverByName('OSM')
data_msk = driver.Open('../data/internal/moscow.osm')
data_peter = driver.Open('../data/internal/peter.osm')

In [41]:
features = []
nlayer = data_msk.GetLayerCount() # 5
print(nlayer)
for i in range(nlayer):
    features += [x for x in data_msk.GetLayerByIndex(i)]
nlayer = data_peter.GetLayerCount() # 5
print(nlayer)
for i in range(nlayer):
    features += [x for x in data_peter.GetLayerByIndex(i)]

5
5


### расстояние до бизнес центров

In [42]:
coords = []
for f in tqdm(features):
    s = str(f.ExportToJson(as_object=True)).lower()
    if 'бизнес' in s and 'центр' in s:
        el = f.ExportToJson(as_object=True)['geometry']['coordinates'][0]
        if type(el) != float:
            coords.append(el)

100%|██████████| 622083/622083 [02:27<00:00, 4215.21it/s]


In [ ]:
# coords = []
# for f in tqdm(features):
#     s = str(f.ExportToJson(as_object=True)).lower()
#     if 'running' in s:
#         coords.append(s)
#         print(s)
#         el = f.ExportToJson(as_object=True)['geometry']['coordinates'][0]
#         if type(el) != float:
#             coords.append(el)

In [43]:
vals1 = df_all[['add_lon', 'add_lat']].drop_duplicates().values.astype(np.float32)
df_vals = pd.DataFrame(vals1, columns=['add_lat', 'add_lon'])
vals2 = np.array(coords, dtype=np.float32)
vals1.shape, vals2.shape

((160184, 2), (206, 2))

In [44]:
X = pairwise_distances(vals1, vals2)
X[X == 0] = 10000

In [45]:
suf = 'bc'
df_vals[suf + '_dist_to'] = X.min(axis=1)
df_vals[suf + '_in_01'] = (X < 0.01).sum(axis=1)
df_vals[suf + '_in_001'] = (X < 0.001).sum(axis=1)
df_vals[suf + '_in_02'] = (X < 0.02).sum(axis=1)
df_vals[suf + '_in_005'] = (X < 0.005).sum(axis=1)
df_vals[suf + '_in_03'] = (X < 0.03).sum(axis=1)

df_all['add_lat_'] = np.round(df_all['add_lat'] * 10000).astype(int) 
df_all['add_lon_'] = np.round(df_all['add_lon'] * 10000).astype(int) 
df_vals['add_lat_'] = np.round(df_vals['add_lat'] * 10000).astype(int) 
df_vals['add_lon_'] = np.round(df_vals['add_lon'] * 10000).astype(int)
del df_vals['add_lat']
del df_vals['add_lon']

df_all = pd.merge(df_all, df_vals, on=['add_lat_', 'add_lon_'], how='left')
del X
del df_all['add_lat_']
del df_all['add_lon_']

### до фастфудов
http://andrewgaidus.com/Convert_OSM_Data/

In [46]:
driver=ogr.GetDriverByName('OSM')
data_msk = driver.Open('../data/internal/moscow.osm')
data_peter = driver.Open('../data/internal/peter.osm')
layer_p = data_msk.GetLayer('points') # 5
features_p = [x for x in layer_p]
layer_p = data_peter.GetLayer('points') # 5
features_p += [x for x in layer_p]

In [47]:
# coords = []
# for f in tqdm(features_p):
#     s = str(f.ExportToJson(as_object=True)).lower()
#     if 'run' in s:
#         print(s)
#         coords.append(f.ExportToJson(as_object=True)['geometry']['coordinates'])

In [48]:
coords = []
for f in tqdm(features_p):
    s = str(f.ExportToJson(as_object=True)).lower()
    if 'fast_food' in s:
        coords.append(f.ExportToJson(as_object=True)['geometry']['coordinates'])

100%|██████████| 343972/343972 [01:16<00:00, 4515.63it/s]


In [49]:
vals1 = df_all[['add_lon', 'add_lat']].drop_duplicates().values.astype(np.float32)
df_vals = pd.DataFrame(vals1, columns=['add_lat', 'add_lon'])
vals2 = np.array(coords, dtype=np.float32)
vals1.shape, vals2.shape

((160184, 2), (2562, 2))

In [50]:
X = pairwise_distances(vals1, vals2)
X[X == 0] = 10000

In [51]:
suf = 'fastfood'
df_vals[suf + '_dist_to'] = X.min(axis=1)
df_vals[suf + '_in_01'] = (X < 0.01).sum(axis=1)
df_vals[suf + '_in_001'] = (X < 0.001).sum(axis=1)
df_vals[suf + '_in_02'] = (X < 0.02).sum(axis=1)
df_vals[suf + '_in_005'] = (X < 0.005).sum(axis=1)
df_vals[suf + '_in_03'] = (X < 0.03).sum(axis=1)

df_all['add_lat_'] = np.round(df_all['add_lat'] * 10000).astype(int) 
df_all['add_lon_'] = np.round(df_all['add_lon'] * 10000).astype(int) 
df_vals['add_lat_'] = np.round(df_vals['add_lat'] * 10000).astype(int) 
df_vals['add_lon_'] = np.round(df_vals['add_lon'] * 10000).astype(int)
del df_vals['add_lat']
del df_vals['add_lon']

df_all = pd.merge(df_all, df_vals, on=['add_lat_', 'add_lon_'], how='left')
del X
del df_all['add_lat_']
del df_all['add_lon_']

## станции

In [52]:
coords = []
for f in tqdm(features_p):
    s = str(f.ExportToJson(as_object=True)).lower()
    if 'railway' in s:
        coords.append(f.ExportToJson(as_object=True)['geometry']['coordinates'])

100%|██████████| 343972/343972 [01:16<00:00, 4482.17it/s]


In [53]:
vals1 = df_all[['add_lon', 'add_lat']].drop_duplicates().values.astype(np.float32)
df_vals = pd.DataFrame(vals1, columns=['add_lat', 'add_lon'])
vals2 = np.array(coords, dtype=np.float32)
vals1.shape, vals2.shape

((160184, 2), (8159, 2))

In [54]:
X = pairwise_distances(vals1, vals2)
X[X == 0] = 10000

In [55]:
suf = 'rail'
df_vals[suf + '_dist_to'] = X.min(axis=1)
df_vals[suf + '_in_01'] = (X < 0.01).sum(axis=1)
df_vals[suf + '_in_001'] = (X < 0.001).sum(axis=1)
df_vals[suf + '_in_02'] = (X < 0.02).sum(axis=1)
df_vals[suf + '_in_005'] = (X < 0.005).sum(axis=1)
df_vals[suf + '_in_03'] = (X < 0.03).sum(axis=1)

df_all['add_lat_'] = np.round(df_all['add_lat'] * 10000).astype(int) 
df_all['add_lon_'] = np.round(df_all['add_lon'] * 10000).astype(int) 
df_vals['add_lat_'] = np.round(df_vals['add_lat'] * 10000).astype(int) 
df_vals['add_lon_'] = np.round(df_vals['add_lon'] * 10000).astype(int)
del df_vals['add_lat']
del df_vals['add_lon']

df_all = pd.merge(df_all, df_vals, on=['add_lat_', 'add_lon_'], how='left')
del X
del df_all['add_lat_']
del df_all['add_lon_']

## райф

In [56]:
coords = []
for f in tqdm(features_p):
    s = str(f.ExportToJson(as_object=True)).lower()
    if 'райф' in s or 'raiffeisen' in s:
        coords.append(f.ExportToJson(as_object=True)['geometry']['coordinates'])

100%|██████████| 343972/343972 [01:15<00:00, 4561.06it/s]


In [57]:
vals1 = df_all[['add_lon', 'add_lat']].drop_duplicates().values.astype(np.float32)
df_vals = pd.DataFrame(vals1, columns=['add_lat', 'add_lon'])
vals2 = np.array(coords, dtype=np.float32)
vals1.shape, vals2.shape

((160184, 2), (194, 2))

In [58]:
X = pairwise_distances(vals1, vals2)
X[X == 0] = 10000

In [59]:
suf = 'raif1'
df_vals[suf + '_dist_to'] = X.min(axis=1)
df_vals[suf + '_in_01'] = (X < 0.01).sum(axis=1)
df_vals[suf + '_in_001'] = (X < 0.001).sum(axis=1)
df_vals[suf + '_in_02'] = (X < 0.02).sum(axis=1)
df_vals[suf + '_in_005'] = (X < 0.005).sum(axis=1)
df_vals[suf + '_in_03'] = (X < 0.03).sum(axis=1)

df_all['add_lat_'] = np.round(df_all['add_lat'] * 10000).astype(int) 
df_all['add_lon_'] = np.round(df_all['add_lon'] * 10000).astype(int) 
df_vals['add_lat_'] = np.round(df_vals['add_lat'] * 10000).astype(int) 
df_vals['add_lon_'] = np.round(df_vals['add_lon'] * 10000).astype(int)
del df_vals['add_lat']
del df_vals['add_lon']

df_all = pd.merge(df_all, df_vals, on=['add_lat_', 'add_lon_'], how='left')
del X
del df_all['add_lat_']
del df_all['add_lon_']

In [60]:
coords = []
for f in tqdm(features):
    s = str(f.ExportToJson(as_object=True)).lower()
    if 'райф' in s or 'raiffeisen' in s:
        el = f.ExportToJson(as_object=True)['geometry']['coordinates'][0]
        if type(el) != float:
            coords.append(el)

100%|██████████| 622083/622083 [02:27<00:00, 4226.28it/s]


In [61]:
vals1 = df_all[['add_lon', 'add_lat']].drop_duplicates().values.astype(np.float32)
df_vals = pd.DataFrame(vals1, columns=['add_lat', 'add_lon'])
vals2 = np.array(coords, dtype=np.float32)
vals1.shape, vals2.shape

((160184, 2), (1, 2))

In [62]:
X = pairwise_distances(vals1, vals2)
X[X == 0] = 10000

In [63]:
suf = 'raif2'
df_vals[suf + '_dist_to'] = X.min(axis=1)
df_vals[suf + '_in_01'] = (X < 0.01).sum(axis=1)
df_vals[suf + '_in_001'] = (X < 0.001).sum(axis=1)
df_vals[suf + '_in_02'] = (X < 0.02).sum(axis=1)
df_vals[suf + '_in_005'] = (X < 0.005).sum(axis=1)
df_vals[suf + '_in_03'] = (X < 0.03).sum(axis=1)

df_all['add_lat_'] = np.round(df_all['add_lat'] * 10000).astype(int) 
df_all['add_lon_'] = np.round(df_all['add_lon'] * 10000).astype(int) 
df_vals['add_lat_'] = np.round(df_vals['add_lat'] * 10000).astype(int) 
df_vals['add_lon_'] = np.round(df_vals['add_lon'] * 10000).astype(int)
del df_vals['add_lat']
del df_vals['add_lon']

df_all = pd.merge(df_all, df_vals, on=['add_lat_', 'add_lon_'], how='left')
del X
del df_all['add_lat_']
del df_all['add_lon_']

In [64]:
del vals2

In [65]:
for c in tqdm(df_all.columns):
    if df_all[c].dtype == np.int64:
        df_all[c] = df_all[c].astype(np.int32)
    if df_all[c].dtype == np.float64:
        df_all[c] = df_all[c].astype(np.float32)

100%|██████████| 699/699 [00:03<00:00, 220.50it/s]


In [67]:
df_all.dtypes.to_csv('../data/df_all_b21_dtypes.csv')
df_all.to_csv('../data/df_all_b21.csv', index=None)

In [66]:
df_all.shape

(2294265, 699)

In [ ]:
%%time
# # Определим типы колонок для экономии памяти
# dtypes = pd.read_csv('../data/df_all_b2_dtypes.csv', header=None, index_col=0).to_dict()[1]
# dtypes.pop('transaction_date', None)
# df_all = pd.read_csv('../data/df_all_b2.csv', dtype=dtypes, parse_dates=['transaction_date'])

# Ранки

In [68]:
gb = df_all.groupby('customer_id')

In [69]:
df_all['rank_amount_cid'] = df_all.groupby('customer_id')['amount'].rank()

In [70]:
df_all = pd.merge(df_all, 
                  df_all.groupby('customer_id')['amount'].agg(['size']).reset_index(), on='customer_id', how='left')

In [71]:
df_all['rank_amount_cid_percent'] = df_all['rank_amount_cid'] / df_all['size']

In [73]:
del features_p

# Расстояния до центров

In [ ]:
am_cols = [c for c in df_all if 'amount' in c]

In [74]:
df_all['dist_to_center'] = -1

In [75]:
vals1 = df_all[df_all.city_name == 'Санкт-Петербург'][['add_lat', 'add_lon']].values
vals2 = np.array([[59.935386, 30.324629]])
X = pairwise_distances(vals1, vals2)
df_all.loc[df_all.city_name == 'Санкт-Петербург', 'dist_to_center'] = X

In [76]:
vals1 = df_all[df_all.city_name == 'Москва'][['add_lat', 'add_lon']].values
vals2 = np.array([[55.7537090, 37.6198133]])
X = pairwise_distances(vals1, vals2)
df_all.loc[df_all.city_name == 'Москва', 'dist_to_center'] = X

# LightGBM

In [77]:
df_all.shape, df_all.columns.duplicated().sum()

((2294265, 703), 0)

In [78]:
df_all = df_all.loc[:,~df_all.columns.duplicated()]

In [79]:
from sklearn.model_selection import train_test_split

ys = ['is_home', 'is_work']
drop_cols = ['atm_address', 'customer_id', 'pos_address', 'terminal_id', 'transaction_date', 
             'is_home' ,'has_home', 'is_work', 'has_work', 'is_train', 'city_name']
drop_cols += ['work_lat','work_lon','home_lat','home_lon', 'string']

drop_cols += ['pred:is_home', 'pred:is_work']
# cols = [c for c in df_all.columns if 'median_dist' in c]
# cols = [c for c in df_all.columns if 'lat' in c or 'lon' in c and 'diff' not in c and 'median' not in c]
# cols += ['address']
# drop_cols += cols

cols = [c for c in df_all.columns if 'mcc_ohe' in c and 'mean' not in c]
# cols += ['address']
drop_cols += cols


y_cols = ['is_home', 'is_work']
usecols = df_all.drop(drop_cols, 1, errors='ignore').columns

In [80]:
params = {
    'objective': 'binary',
    'num_leaves': 511,
    'learning_rate': 0.01,
    'metric' : 'binary_logloss',
    'feature_fraction': 0.8,
    'bagging_fraction': 0.8,
    'bagging_freq': 1,
    'num_threads': 12,
    'verbose': 0,
}

model = {}

In [81]:
y_col = 'is_home'

cust_train = df_all[df_all['is_train']==1].groupby('customer_id')[y_col.replace('is_','has_')].max()
cust_train = cust_train[cust_train > 0].index

cust_train, cust_valid = train_test_split(cust_train, test_size=0.2, shuffle=True, random_state=111)

df_train = pd.DataFrame(cust_train, columns=['customer_id']).merge(df_all, how='left')
df_valid = pd.DataFrame(cust_valid, columns=['customer_id']).merge(df_all, how='left')

lgb_train = lgb.Dataset(df_train[usecols], df_train[y_col])
lgb_valid = lgb.Dataset(df_valid[usecols], df_valid[y_col])

gbm_h = lgb.train(params,
                lgb_train,
                valid_sets=[lgb_valid],
                num_boost_round=2000,
                verbose_eval=30,
                early_stopping_rounds=100)

model[y_col] = gbm_h

Training until validation scores don't improve for 100 rounds.
[30]	valid_0's binary_logloss: 0.583925
[60]	valid_0's binary_logloss: 0.516789
[90]	valid_0's binary_logloss: 0.47214
[120]	valid_0's binary_logloss: 0.442272
[150]	valid_0's binary_logloss: 0.421676
[180]	valid_0's binary_logloss: 0.40754
[210]	valid_0's binary_logloss: 0.397597
[240]	valid_0's binary_logloss: 0.390218
[270]	valid_0's binary_logloss: 0.385013
[300]	valid_0's binary_logloss: 0.381261
[330]	valid_0's binary_logloss: 0.378859
[360]	valid_0's binary_logloss: 0.376952
[390]	valid_0's binary_logloss: 0.375839
[420]	valid_0's binary_logloss: 0.375117
[450]	valid_0's binary_logloss: 0.3746
[480]	valid_0's binary_logloss: 0.374571
[510]	valid_0's binary_logloss: 0.374397
[540]	valid_0's binary_logloss: 0.374263
[570]	valid_0's binary_logloss: 0.374408
[600]	valid_0's binary_logloss: 0.374786
[630]	valid_0's binary_logloss: 0.375561
Early stopping, best iteration is:
[556]	valid_0's binary_logloss: 0.374234


In [82]:
y_col = 'is_work'

cust_train = df_all[df_all['is_train']==1].groupby('customer_id')[y_col.replace('is_','has_')].max()
cust_train = cust_train[cust_train > 0].index

cust_train, cust_valid = train_test_split(cust_train, test_size=0.2, shuffle=True, random_state=111)



df_train = pd.DataFrame(cust_train, columns=['customer_id']).merge(df_all, how='left')
df_valid = pd.DataFrame(cust_valid, columns=['customer_id']).merge(df_all, how='left')

lgb_train = lgb.Dataset(df_train[usecols], df_train[y_col])
lgb_valid = lgb.Dataset(df_valid[usecols], df_valid[y_col])

gbm_w = lgb.train(params,
                lgb_train,
                valid_sets=[lgb_valid],
                num_boost_round=2000,
                verbose_eval=30,
                early_stopping_rounds=100)

model[y_col] = gbm_w

Training until validation scores don't improve for 100 rounds.
[30]	valid_0's binary_logloss: 0.561026
[60]	valid_0's binary_logloss: 0.481483
[90]	valid_0's binary_logloss: 0.430032
[120]	valid_0's binary_logloss: 0.396466
[150]	valid_0's binary_logloss: 0.374586
[180]	valid_0's binary_logloss: 0.359539
[210]	valid_0's binary_logloss: 0.349693
[240]	valid_0's binary_logloss: 0.343236
[270]	valid_0's binary_logloss: 0.339659
[300]	valid_0's binary_logloss: 0.337191
[330]	valid_0's binary_logloss: 0.336102
[360]	valid_0's binary_logloss: 0.33665
[390]	valid_0's binary_logloss: 0.337185
[420]	valid_0's binary_logloss: 0.339059
Early stopping, best iteration is:
[336]	valid_0's binary_logloss: 0.336068


In [ ]:
gbm_w

## Полезные MCC
дом
6011 - финансы
5411 - придомовые магазы
5814 - мак
5912 - аптеки
5921 - пиво
5499 - магазы пяторочка типа
5812 - рестроанчики
работа


In [ ]:
figsize(14, 10)
lgb.plot_importance(gbm_h, max_num_features=40)

In [83]:
def _best(x):
    ret = None
    for col in ys:
        pred = ('pred:%s' % col)
        if pred in x:
            i = (x[pred].idxmax())
            cols = [pred, 'add_lat', 'add_lon']
            if col in x:
                cols.append(col)
            tmp = x.loc[i,cols]
            tmp.rename({
                'add_lat':'%s:add_lat' % col,
                'add_lon':'%s:add_lon' % col,
            }, inplace = True)
            if ret is None:
                ret = tmp
            else:
                ret = pd.concat([ret, tmp])
    return ret


def predict_proba(dt, ys=['is_home', 'is_work']):
    for col in ys:
        pred = ('pred:%s' % col)
        dt[pred] = model[col].predict(dt[usecols])
    return dt.groupby('customer_id').apply(_best).reset_index()

def score(dt, ys=['is_home', 'is_work'], return_df=False):
    dt_ret = predict_proba(dt, ys)
    if return_df:
        return dt_ret
    mean = 0.0
    for col in ys:
        col_mean = dt_ret[col].mean()
        mean += col_mean
    if len(ys) == 2:
        mean = mean / len(ys)
    return mean

In [84]:
print ("Train accuracy:", score(df_train, ys=['is_home']))
print ("Test accuracy:", score(df_valid, ys=['is_home']))

print ("Train accuracy:", score(df_train, ys=['is_work']))
print ("Test accuracy:", score(df_valid, ys=['is_work']))

Train accuracy: 0.640571982549685
Test accuracy: 0.6443798449612403
Train accuracy: 0.5368395540475036
Test accuracy: 0.3536821705426357


Train accuracy: 0.5458070770722249
Test accuracy: 0.5494186046511628
Train accuracy: 0.4301987396994668
Test accuracy: 0.3536821705426357

# Анализ False-Negative

In [ ]:
# сколько вообще людей имеют хорошую точку
df_all[(df_all.is_train == 1)].groupby('customer_id')['is_work'].agg('max').mean()

In [ ]:
df_pred = score(df_valid, ys=['is_home'], return_df=True)

In [ ]:
df_pred.sample(5)

In [ ]:
cid = 'bf66305d0ec05abb6e6a6358acb8c2a1'
cid = df_pred[df_pred.is_home == 0].sample(1)['customer_id'].values[0]

df_an = df_all[df_all.customer_id == cid]
center_home = df_an[['home_lat', 'home_lon']].drop_duplicates().values
center_work = df_an[['work_lat', 'work_lon']].drop_duplicates().values


predicted_home = df_pred[df_pred.customer_id == cid][['is_home:add_lat', 'is_home:add_lon']].drop_duplicates().values
predicted_work = df_pred[df_pred.customer_id == cid][['is_work:add_lat', 'is_work:add_lon']].drop_duplicates().values

points_pos = df_an[df_an.is_pos == 1][['add_lat', 'add_lon']].dropna().values
points_atm = df_an[df_an.is_pos == 0][['add_lat', 'add_lon']].dropna().values
print(center_home.shape, center_work.shape, points_pos.shape, points_atm.shape)

# синие - покупки
# красные - банкоматы
gmap = gmaps.Map()
if len(points_pos) > 0:
    gmap.add_layer(gmaps.symbol_layer(points_pos, hover_text='pos', 
                                      fill_color="blue", stroke_color="blue", scale=3))
if len(points_atm) > 0:
    gmap.add_layer(gmaps.symbol_layer(points_atm, hover_text='atm',
                                      fill_color="red", stroke_color="red",scale=3))

if not np.isnan(center_home)[0][0]:
    gmap.add_layer(gmaps.marker_layer(center_home, label='home'))
if not np.isnan(center_work)[0][0]:
    gmap.add_layer(gmaps.marker_layer(center_work, label='work'))

gmap.add_layer(gmaps.marker_layer(predicted_home, label='predicted_home'))
gmap.add_layer(gmaps.marker_layer(predicted_work, label='predicted_work'))
    
gmap

In [ ]:
df_all.to_csv('../data/dfpredict1903.csv', index=None)

# Predict

In [ ]:
del cust_test

In [85]:
cust_test = df_all.loc[df_all['is_train'] == 0, 'customer_id'].unique()
# df_test = pd.DataFrame(cust_test, columns = ['customer_id']).merge(df_all, how = 'left')
df_test = predict_proba(pd.DataFrame(cust_test, columns = ['customer_id']).merge(df_all, how = 'left'))
df_test.rename(columns = {
        'customer_id':'_ID_',
        'is_home:add_lat': '_HOME_LAT_',
        'is_home:add_lon': '_HOME_LON_',
        'is_work:add_lat': '_WORK_LAT_',
        'is_work:add_lon': '_WORK_LON_'}, inplace = True)
df_test = df_test[['_ID_', '_WORK_LAT_', '_WORK_LON_', '_HOME_LAT_', '_HOME_LON_']]

df_test.head()

_ID_  _WORK_LAT_  _WORK_LON_  _HOME_LAT_  \
0  00021683ccb416637fe9a4cd35e4606e  55.026001   82.915001   55.041073    
1  0002d0f8a642272b41c292c12ab6e602  44.033001   42.835999   44.036594    
2  0004d182d9fede3ba2534b2d5e5ad27e  43.585999   39.723999   43.572186    
3  0008c2445518c9392cb356c5c3db3392  51.530449   46.033218   51.533936    
4  000b373cc4969c0be8e0933c08da67e1  56.319836   43.925976   56.247688    

   _HOME_LON_  
0  82.980629   
1  42.855629   
2  39.734737   
3  46.025490   
4  43.463734

# Формируем submission-файл

In [86]:
# Заполняем пропуски
df_ = pd.read_csv('../data/test_set.csv', dtype=dtypes, usecols=['customer_id'])
submission = pd.DataFrame(df_['customer_id'].unique(), columns=['_ID_'])

submission = submission.merge(df_test, how='left').fillna(0)
# Пишем файл submission
submission.to_csv('../submissions/base_16_644_353.csv', index=None)

In [ ]:
submission_2 = pd.read_csv('../submissions/base_11_625_34.csv')

In [ ]:
submission.head()

In [ ]:
submission_2.head()

In [ ]:
submission_3 = submission_2.copy()
submission_3['_WORK_LAT_'] = (submission['_WORK_LAT_'] + submission_2['_WORK_LAT_'])/2
submission_3['_WORK_LON_'] = (submission['_WORK_LON_'] + submission_2['_WORK_LON_'])/2
submission_3['_HOME_LAT_'] = (submission['_HOME_LAT_'] + submission_2['_HOME_LAT_'])/2
submission_3['_HOME_LON_'] = (submission['_HOME_LON_'] + submission_2['_HOME_LON_'])/2
submission_3.to_csv('../submissions/base_15and12/2.csv', index=None)